crates/typed-store-macros/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+
+        impl DBMapTableUtil for #name {
+            /// Opens a set of tables in read-write mode
+            /// Only one process is allowed to do this at a time
+            fn open_tables_read_write(
+                path: PathBuf,
+                db_options: Option<RocksDBOptions>,
+            ) -> Self {
+                Self::open_tables_impl(path, None, db_options)
+            }
+
+            /// Open in read only mode. No limitation on number of processes to do this
+            fn open_tables_read_only(
+                path: PathBuf,
+                with_secondary_path: Option<PathBuf>,
+                db_options: Option<RocksDBOptions>,
+            ) -> Self {
+                match with_secondary_path {
+                    Some(q) => Self::open_tables_impl(path, Some(q), db_options),
+                    None => {
+                        let p: PathBuf = tempfile::tempdir()
+                        .expect("Failed to open temporary directory")
+                        .into_path();
+                        Self::open_tables_impl(path, Some(p), db_options)
+                    }
+                }
+            }
+
+            /// Opens a set of tables in read-write mode
+            /// If with_secondary_path is set, the DB is opened in read only mode with the path specified
+            fn open_tables_impl(
+                path: PathBuf,
+                with_secondary_path: Option<PathBuf>,
+                db_options: Option<RocksDBOptions>,
+            ) -> Self {
+                let path = &path;
+                let db_options = db_options.unwrap_or_default().clone();
+
+                let db = {
+                    let opt_cfs: &[(&str, &RocksDBOptions)] = &[
+                        #(
+                            (stringify!(#field_names), &Self::adjusted_db_options(None, #cache_capacity, #point_lookup).clone()),
+                        )*
+                    ];
+
+                    if let Some(p) = with_secondary_path {
+                        typed_store::rocks::open_cf_opts_secondary(path, Some(&p), Some(db_options), opt_cfs)
+                    } else {
+                        typed_store::rocks::open_cf_opts(path, Some(db_options), opt_cfs)
+                    }
+                }.expect("Cannot open DB.");
+
+                let (
+                        #(
+                            #field_names,
+                        )*
+                ) = (#(
+                        DBMap::#inner_types::reopen(&db, Some(stringify!(#field_names))).expect(&format!("Cannot open {} CF.", stringify!(#field_names))[..])
+                    ), *);
+
+                Self {
+                    #(
+                        #field_names,
+                    )*
+                }
+            }
+
+            /// Given a provided `db_options`, add a few default options.
+            /// Returns the default option and the point lookup option.
+            fn adjusted_db_options(
```

Comment:
looks like this function don't really need to be in the macro? Can it live in the trait instead?